README.md
# HM Redirects

Allows to redirect one path to another path on the same domain.

## Architecture
Redirects are stored as a custom post type and use the following fields:

- `post_name` to store the md5 hash of the _From_ path. This column is used because it is indexed, and allows fast queries. `md5` is used to simplify the storage.
- `post_title` to store the _From_ path.
- `post_excerpt`to store the the _To_ path.

## Tips
This plugin uses `wp_safe_redirect()` to redirect. You will have to whitelist your redirect target domains using WordPress' `allowed_redirect_hosts` filter, otherwise the redirect will not work.
One way to get a list of redirect target domains is to run the WP-CLI command: `wp hm-redirects find-domains`. Another is to add them dynamically just-in-time using the filter `hm_redirects_matched_redirect`.

## Attributions
Props for the data storage approach to VIP's [WPCOM Legacy Redirector](https://github.com/Automattic/WPCOM-Legacy-Redirector).


changelog.txt
== Changelog ==

= 0.5.2 =
* Save URLs as provided rather than lowercasing them #38

= 0.5.1 =
* Fix no trailing slash bug #34

= 0.5.0 =
* Fix redirect CPT to be private #26
* Add filters to modify the contents of the HTTP Status Code drop down #21
* Sanitise full from URL to relative path #16
* Fix redirects with query arguments bug #15
* Remove title field from Post Edit screen #14
* Improve Redirect listing screen #13
* Increase the size of the From and To URL fields #12
* Add a filter to modify URLs before they are being saved #11
* Set the order of the menu item to sit next to Tools #6
* Hide Visibility and Date Settings in the Post Box #1
* Fix Redirect Matching with non ASCII URLs #24

= 0.4.2 =
* Fix fatal error during WP CLI script execution caused by incorrect class name when calling WP_CLI::add_command
* Fix php linting errors

= 0.4.1 =
* Use `get_page_by_path()` instead of a `WP_Query` in `get_redirect_post()`. The `get_page_by_path()` function uses
  an object cache bucket for post paths to id matches, therefore potentially speeding up the lookup.

= 0.4 =
* Remove usage of `filter_var()` to support domains with non-ASCII characters.

= 0.3 =
* Add WP-CLI commands.

= 0.2 =
* Internationalisation improvements.

= 0.1 =
* Initial release. Handles redirects path and query argument redirects on the same domain.


tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Hm_Redirects
 */

$_tests_dir = getenv( 'WP_TESTS_DIR' );

if ( ! $_tests_dir ) {
	$_tests_dir = rtrim( sys_get_temp_dir(), '/\\' ) . '/wordpress-tests-lib';
}

if ( ! file_exists( $_tests_dir . '/includes/functions.php' ) ) {
	throw new Exception( "Could not find $_tests_dir/includes/functions.php, have you run bin/install-wp-tests.sh ?" );
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/hm-redirects.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/includes/bootstrap.php';


tests/atest-utils.php
<?php
/**
 * The Sun Redirector.
 *
 * @package TheSun_Redirector_Plugin
 * @author Human Made Limited
 */

/**
 * Class Utils_Test
 */
class Utils_Test extends TheSun_TestCase {

	/**
	 * @var TheSun_Redirector_Utils
	 */
	protected $utils;

	/**
	 *
	 */
	public function setUp() {
		$this->utils = new TheSun_Redirector_Utils();
	}

	/**
	 * Provides valid test URLS.
	 *
	 * @return array
	 */
	public function provider_normalised_url_valid() {
		return [
			[ '/only/the/path', '/only/the/path' ],
			[ '/a/path/?with=queryparam', '/a/path/?with=queryparam' ],
			[ '/a/path/withbrakets?foo[bar]=baz', '/a/path/withbrakets?foo[bar]=baz' ],
			[ '/Mr%20%20WordPress', '/Mr%20%20WordPress' ],
		];
	}

	/**
	 * Provides invalid test URLs.
	 *
	 * @return array
	 */
	public function provider_normalised_url_invalid() {
		return [
			[ 'http://thesun.dev' ], // No path.
			[ 'http://url%20invalid%20charcaters' ], // Invalid character.
			[ '/invalidcharacters¡™£¢∞§¶•ªº%5B%5D/here' ], // Invalid character.
		];
	}

	/**
	 * Provides valid test URLS.
  *
	 * @return array
	 */
	public function provider_redirect_uri_valid() {
		return [
			[ '/original-post', '/redirected-post', 301 ],
			[ '/original-post?with=query-param', '/redirected-post?with=query-param', 303 ],
		];
	}

	/**
	 * Tests normalise_url.
	 *
	 * @dataProvider provider_normalised_url_valid
	 *
	 * @param string $original_url Original URL.
	 * @param string $expected_result Expected result.
	 */
	public function test_normalise_url_returns_normalized_url( $original_url, $expected_result ) {
		$result = $this->utils->normalise_url( $original_url );
		$this->assertEquals( $expected_result, $result );
	}

	/**
	 * Tests normalise_url.
	 *
	 * @dataProvider provider_normalised_url_invalid
	 *
	 * @param string $original_url Original URL.
	 */
	public function test_normalize_url_returns_error_for_invalid_urls( $original_url ) {
		$result = $this->utils->normalise_url( $original_url );
		$this->assertInstanceOf( 'WP_Error', $result );
	}

	/**
	 * Tests get_redirect_post.
	 */
	public function test_get_redirect_post_invalid_urls() {
		$this->assertNull( $this->invokeMethod( $this->utils, 'get_redirect_post', array( 'sfsgsdfgdfgdfgdf' ) ) );
	}

	/**
	 * Tests get_redirect_post.
	 */
	public function test_get_redirect_post_valid_urls() {
		$p = $this->factory->post->create( array( 'post_title' => 'Test Post', 'post_name' => md5( '/test/this/path' ), 'post_type' => 'thesun-redirects' ) );
		$this->assertInstanceOf( 'WP_Post', $this->invokeMethod( $this->utils, 'get_redirect_post', array( '/test/this/path' ) ) );

		wp_delete_post( $p );
	}

	/**
	 * Tests get_redirect_uri.
	 *
	 * @dataProvider provider_redirect_uri_valid
	 *
	 * @param string $original_url Original URL.
	 * @param string $expected_result Expected result.
	 */
	public function test_get_redirect_uri_valid_urls( $original_url, $expected_result ) {

		$p = $this->factory->post->create(
			[
				'post_title'   => 'Test Post',
				'post_name'    => md5( $original_url ),
				'post_type'    => 'thesun-redirects',
				'post_excerpt' => $expected_result,
			]
		);

		$result = $this->utils->get_redirect_uri( $original_url );
		$this->assertEquals( home_url() . $expected_result, $result );

		wp_delete_post( $p );
	}

	/**
	 * Tests get_redirect_status_code.
	 *
	 * @dataProvider provider_redirect_uri_valid
	 *
	 * @param string $original_url Original URL.
	 * @param string $expected_result Expected result.
	 * @param int    $status_code HTTP status code.
	 */
	public function test_get_redirect_status_code( $original_url, $expected_result, $status_code ) {
		// make sure we catch error when argument is missing.
		$this->assertInstanceOf( 'WP_Error', $this->utils->get_redirect_status_code() );

		$this->assertEquals( 302, $this->utils->get_redirect_status_code( '"^<>{}`' ) );

		$p = $this->factory->post->create(
			[
				'post_title'            => 'Test Post',
				'post_name'             => md5( $original_url ),
				'post_type'             => 'thesun-redirects',
				'post_excerpt'          => $expected_result,
				'post_content_filtered' => $status_code
			]
		);

		$this->assertEquals( $status_code, $this->utils->get_redirect_status_code( $original_url ) );

		wp_delete_post( $p );
	}
}


tests/test-admin-ui.php
<?php
/**
 * Test for the admin UI.
 */

namespace HM\Redirects\Tests;

use HM\Redirects\Admin_UI;
use WP_UnitTestCase;

class Admin_UI_Test extends WP_UnitTestCase {

	/**
	 * Test the validation method.
	 */
	public function test_validate_meta() {
		$this->assertSame( 'Fields are required', Admin_UI\validate_meta( '', '' ) );
	}
}


includes/cli.php
<?php
/**
 * WP-CLI extension for HM Redirects.
 *
 * Originally taken from Automattic's WPCOM-Legacy-Redirector plugin. ❤️
 * https://github.com/Automattic/WPCOM-Legacy-Redirector
 *
 * For familiarity reasons, these commands work the same as those in that plugin.
 * However our "redirect from" URLs need to be relative to the root of the site,
 * and have a leading slash.
 *
 * @package hm-redirects
 */

namespace HM\Redirects\CLI;

use WP_CLI, WP_CLI_Command;
use HM\Redirects\{Utilities, Handle_Redirects};
use const HM\Redirects\Post_Type\SLUG as REDIRECTS_POST_TYPE;

/**
 * Handles redirects in a scalable manner.
 */
class Commands extends WP_CLI_Command {

	/**
	 * Find domains redirected to.
	 *
	 * This is useful to populate WordPress' `allowed_redirect_hosts` filter.
	 *
	 * @subcommand find-domains
	 *
	 * @param string[] $args Not used.
	 * @param string[] $assoc_args Not used.
	 */
	public function find_domains( array $args, array $assoc_args ) {
		global $wpdb;

		$domains        = array();
		$paged          = 0;
		$posts_per_page = 500;

		$total_redirects = (int) $wpdb->get_var(
			$wpdb->prepare(
				"SELECT COUNT( ID ) FROM $wpdb->posts WHERE post_type = %s AND post_excerpt LIKE %s",
				REDIRECTS_POST_TYPE,
				'http%'
			)
		);

		$progress_bar = WP_CLI\Utils\make_progress_bar( 'Finding domains', $total_redirects );

		do {
			$redirect_urls = $wpdb->get_col(
				$wpdb->prepare(
					"SELECT post_excerpt FROM $wpdb->posts WHERE post_type = %s AND post_excerpt LIKE %s ORDER BY ID ASC LIMIT %d, %d",
					REDIRECTS_POST_TYPE,
					'http%',
					( $paged * $posts_per_page ),
					$posts_per_page
				)
			);

			foreach ( $redirect_urls as $redirect_url ) {
				$progress_bar->tick();

				if ( ! empty( $redirect_url ) ) {
					$redirect_host = parse_url( $redirect_url, PHP_URL_HOST );
					if ( $redirect_host ) {
						$domains[] = $redirect_host;
					}
				}
			}

			// Throttle.
			sleep( 1 );
			$paged++;
		} while ( count( $redirect_urls ) );

		$progress_bar->finish();

		$domains = array_unique( $domains );
		WP_CLI::line( sprintf( 'Found %s unique outbound domains', number_format( count( $domains ) ) ) );

		foreach ( $domains as $domain ) {
			WP_CLI::line( $domain );
		}
	}

	/**
	 * Insert a single redirect
	 *
	 * from_url_relative must be relative to the root of the site.
	 *
	 * @subcommand insert-redirect
	 * @synopsis <from_url_relative> <to_url_absolute>
	 *
	 * @param string[] $args Positional arguments.
	 * @param string[] $assoc_args Not used.
	 */
	public function insert_redirect( $args, $assoc_args ) {
		$from = $args[0];

		if ( ctype_digit( $args[1] ) ) {
			$to = get_permalink( (int) $args[1] );
			if ( ! $to ) {
				WP_CLI::error( sprintf( 'Destination post %s cannot be found', $args[1] ) );
			}

		} else {
			$to = esc_url_raw( $args[1] );
		}

		$redirect = Utilities\insert_redirect( $from, $to, 301 );

		if ( is_wp_error( $redirect ) ) {
			WP_CLI::error( sprintf(
				"Couldn't insert %s -> %s: %s",
				$from,
				$to,
				implode( PHP_EOL, $redirect->get_error_messages() )
			) );
		}

		WP_CLI::success( sprintf( "Inserted %s -> %s", $from, $to ) );
	}

	/**
	 * Bulk import redirects from a CSV file matching the following structure:
	 *
	 * redirect_from_path,(redirect_to_post_id|redirect_to_path|redirect_to_url)
	 *
	 * ## OPTIONS
	 *
	 * [--format=<format>]
	 * : Render output in a particular format.
	 * ---
	 * default: csv
	 * options:
	 *   - table
	 *   - json
	 *   - yaml
	 *   - csv
	 * ---
	 *
	 * [--verbose]
	 *
	 * @subcommand import-from-csv
	 * @synopsis --csv=<path-to-csv> [--format=<format>] [--verbose]
	 *
	 * @param string[] $args Positional arguments.
	 * @param string[] $assoc_args Associative arguments.
	 */
	public function import_from_csv( $args, $assoc_args ) {
		define( 'WP_IMPORTING', true );

		$csv     = trim( WP_CLI\Utils\get_flag_value( $assoc_args, 'csv' ) );
		$format  = WP_CLI\Utils\get_flag_value( $assoc_args, 'format' );
		$verbose = isset( $assoc_args['verbose'] );

		if ( ! $csv || ! file_exists( $csv ) ) {
			WP_CLI::error( "Invalid 'csv' file" );
		}

		if ( ! $verbose ) {
			WP_CLI::line( 'Processing...' );
		}

		$handle = fopen( $csv, 'r' );
		if ( $handle === false ) {
			WP_CLI::error( "Cannot open 'csv' file" );
		}

		$notices = array();
		$row     = 0;

		while ( ( $data = fgetcsv( $handle, 2000, "," ) ) !== false ) {
			$row++;
			$redirect_from = $data[ 0 ];
			$redirect_to   = $data[ 1 ];

			// Convert "redirect to" post IDs to permalinks.
			if ( ctype_digit( $redirect_to ) ) {
				$redirect_to = get_permalink( (int) $redirect_to );

				if ( ! $redirect_to ) {
					$notices[] = array(
						'redirect_from' => $redirect_from,
						'redirect_to'   => $data[ 1 ],
						'message'       => 'Skipped - could not find redirect_to post',
					);

					continue;
				}
			}

			if ( $verbose ) {
				WP_CLI::line( "Adding (CSV) redirect for {$redirect_from} to {$redirect_to}" );
				WP_CLI::line( "-- at $row" );
			} elseif ( 0 === $row % 100 ) {
				WP_CLI::line( "Processing row $row" );
			}

			$redirect = Utilities\insert_redirect(
				$redirect_from,
				esc_url_raw( $redirect_to ),
				301
			);

			// Record any error notices.
			if ( is_wp_error( $redirect ) ) {
				$notices[] = array(
					'redirect_from' => $redirect_from,
					'redirect_to'   => $redirect_to,
					'message'       => sprintf(
						'Could not insert redirect: %s',
						implode( PHP_EOL, $redirect->get_error_messages() )
					),
				);

			// Record success notices.
			} elseif ( $verbose ) {
				$notices[] = array(
					'redirect_from' => $redirect_from,
					'redirect_to'   => $redirect_to,
					'message'       => 'Successfully imported',
				);
			}

			if ( 0 === $row % 100 ) {
				Utilities\clear_object_cache();

				// Throttle writes.
				sleep( 1 );
			}
		}

		fclose( $handle );

		if ( count( $notices ) > 0 ) {
			WP_CLI\Utils\format_items( $format, $notices, array( 'redirect_from', 'redirect_to', 'message' ) );
		} else {
			echo WP_CLI::colorize( "%GAll of your redirects have been imported. Nice work!%n " );
		}
	}

	/**
	 * Bulk import redirects from URLs stored as meta values for posts.
	 *
	 * ## OPTIONS
	 *
	 * [--start=<start-offset>]
	 *
	 * [--end=<end-offset>]
	 *
	 * [--skip_dupes=<skip-dupes>]
	 *
	 * [--format=<format>]
	 * : Render output in a particular format.
	 * ---
	 * default: csv
	 * options:
	 *   - table
	 *   - json
	 *   - yaml
	 *   - csv
	 * ---
	 *
	 * [--dry_run]
	 *
	 * [--verbose]
	 * : Display notices for sucessful imports and duplicates (if skip_dupes is used)
	 *
	 * @subcommand import-from-meta
	 * @synopsis --meta_key=<name-of-meta-key> [--start=<start-offset>] [--end=<end-offset>] [--skip_dupes=<skip-dupes>] [--format=<format>] [--dry_run] [--verbose]
	 *
	 * @param string[] $args Positional arguments.
	 * @param string[] $assoc_args Associative arguments.
	 */
	function import_from_meta( $args, $assoc_args ) {
		global $wpdb;

		define( 'WP_IMPORTING', true );

		$meta_key   = isset( $assoc_args['meta_key'] ) ? sanitize_key( $assoc_args['meta_key'] ) : 'change-me';
		$offset     = isset( $assoc_args['start'] ) ? intval( $assoc_args['start'] ) : 0;
		$end_offset = isset( $assoc_args['end'] ) ? intval( $assoc_args['end'] ) : 99999999;;

		$skip_dupes = isset( $assoc_args['skip_dupes'] ) ? (bool) intval( $assoc_args['skip_dupes'] ) : false;
		$dry_run    = isset( $assoc_args['dry_run'] );
		$format     = WP_CLI\Utils\get_flag_value( $assoc_args, 'format' );
		$verbose    = isset( $assoc_args['verbose'] );

		if ( $dry_run ) {
			WP_CLI::line( '---Dry Run---' );
		} else {
			WP_CLI::line( '---Live Run--' );
		}

		// Check we have any work to do.
		$total_redirects = $wpdb->get_var(
			$wpdb->prepare(
				"SELECT COUNT( post_id ) FROM $wpdb->postmeta WHERE meta_key = %s",
				$meta_key
			)
		);

		if ( $total_redirects === 0 ) {
			WP_CLI::error( sprintf( 'No redirects found for meta_key: %s', $meta_key ) );
		}

		$progress_bar = WP_CLI\Utils\make_progress_bar( sprintf( 'Importing %s redirects', number_format( $total_redirects ) ), $total_redirects );
		$notices      = array();

		// Start the import; loop through batches of posts.
		do {
			$redirects = $wpdb->get_results(
				$wpdb->prepare(
					"SELECT post_id as redirect_to_post_id, meta_value as abs_redirect_from FROM $wpdb->postmeta WHERE meta_key = %s ORDER BY post_id ASC LIMIT %d, 1000",
					$meta_key,
					$offset
				)
			);

			$total = count( $redirects );
			$row   = 0;

			foreach ( $redirects as $redirect ) {
				$row++;
				$progress_bar->tick();

				// "Redirect from" parameter must be relative to the root of the site.
				$redirect_from = wp_parse_url( $redirect->abs_redirect_from, PHP_URL_PATH );
				$query_args    = wp_parse_url( $redirect->abs_redirect_from, PHP_URL_QUERY );

				if ( $query_args !== null ) {
					$redirect_from = "?{$query_args}";
				}

				// The "redirect to" value is a post ID. Grab the appropriate URL.
				$redirect_to = get_permalink( (int) $redirect->redirect_to_post_id );
				if ( ! $redirect_to ) {
					$notices[] = array(
						'redirect_from' => $redirect_from,
						'redirect_to'   => $redirect->redirect_to_post_id,
						'message'       => 'Skipped - could not find redirect_to post',
					);

					continue;
				}


				if ( $skip_dupes ) {
					$has_existing_redirect = Handle_Redirects\get_redirect_post( $redirect_from );
					$has_existing_redirect = ( $has_existing_redirect === null ) ? false : true;

					if ( $has_existing_redirect ) {
						if ( $verbose ) {
							$notices[] = array(
								'redirect_from' => $redirect_from,
								'redirect_to'   => $redirect_to,
								'message'       => sprintf( 'Skipped - "redirect from" URL already exists (%s)', $redirect_from ),
							);
						}

						continue;
					}
				}

				// Add redirects.
				if ( $dry_run === false ) {
					$redirect = Utilities\insert_redirect(
						$redirect_from,
						$redirect_to,
						301
					);

					// Record any error notices.
					if ( is_wp_error( $redirect ) ) {
						$notices[] = array(
							'redirect_from' => $redirect_from,
							'redirect_to'   => $redirect_to,
							'message'       => sprintf(
								'Could not insert redirect: %s',
								implode( PHP_EOL, $redirect->get_error_messages() )
							),
						);

					// Record success notices.
					} elseif ( $verbose ) {
						$notices[] = array(
							'redirect_from' => $redirect_from,
							'redirect_to'   => $redirect_to,
							'message'       => 'Successfully imported',
						);
					}
				}

				if ( 0 === $row % 100 ) {
					Utilities\clear_object_cache();

					// Throttle writes.
					sleep( 1 );
				}
			}

			$offset += 1000;
		} while( $total >= 1000 && $offset < $end_offset );

		$progress_bar->finish();

		if ( count( $notices ) > 0 ) {
			WP_CLI\Utils\format_items( $format, $notices, array( 'redirect_from', 'redirect_to', 'message' ) );
		} else {
			echo WP_CLI::colorize( "%GAll of your redirects have been imported. Nice work!%n " );
		}
	}
}


includes/post-type.php
<?php
/**
 * Register the redirects custom post type.
 *
 * @package hm-redirects
 */

namespace HM\Redirects\Post_Type;

/**
 * Redirects post type slug.
 */
const SLUG = 'hm_redirect';

/**
 * Setup hooks.
 */
function setup() {
	add_action( 'init', __NAMESPACE__ . '\\register_post_type' );
}

/**
 * Register the post type.
 */
function register_post_type() {
	$labels = [
		'name'               => esc_html__( 'Redirects', 'hm-redirects' ),
		'singular_name'      => esc_html__( 'Redirect', 'hm-redirects' ),
		'add_new'            => esc_html__( 'Add New', 'hm-redirects' ),
		'add_new_item'       => esc_html__( 'Add New Redirect', 'hm-redirects' ),
		'edit_item'          => esc_html__( 'Edit Redirect', 'hm-redirects' ),
		'new_item'           => esc_html__( 'New Redirect', 'hm-redirects' ),
		'view_item'          => esc_html__( 'View Redirect', 'hm-redirects' ),
		'search_items'       => esc_html__( 'Search Redirects', 'hm-redirects' ),
		'not_found'          => esc_html__( 'No redirects found', 'hm-redirects' ),
		'not_found_in_trash' => esc_html__( 'No redirects found in trash', 'hm-redirects' ),
		'all_items'          => esc_html__( 'All Redirects', 'hm-redirects' ),
	];

	\register_post_type(
		SLUG,
		[
			'labels'              => $labels,
			'show_in_feed'        => false,
			'supports'            => [ 'title' ],
			'hierarchical'        => false,
			'public'              => false,
			'show_ui'             => true,
			'show_in_menu'        => true,
			'menu_position'       => 75,
			'menu_icon'           => 'dashicons-migrate',
			'show_in_admin_bar'   => true,
			'show_in_nav_menus'   => false,
			'can_export'          => true,
			'has_archive'         => false,
			'exclude_from_search' => true,
			'publicly_queryable'  => false,
		]
	);

	remove_post_type_support( SLUG, 'title' );
}


includes/handle-redirects.php
<?php
/**
 * Handle redirecting URLs if a redirect has been set.
 *
 * @package hm-redirects
 */

namespace HM\Redirects\Handle_Redirects;

use HM\Redirects\Post_Type as Redirects_Post_Type;
use HM\Redirects\Utilities;
use WP_Error;

/**
 * Initialise the plugin.
 */
function setup() {
	add_action( 'parse_request', __NAMESPACE__ . '\\maybe_do_redirect', 0 );
}

/**
 * Determine if a redirect needs to be performed for this URL, and do it if so.
 */
function maybe_do_redirect() {
	if ( is_admin() ) {
		return false;
	}

	if ( ! isset( $_SERVER['REQUEST_URI'] ) ) {
		return false;
	}

	$path = untrailingslashit( str_replace( wp_parse_url( home_url(), PHP_URL_PATH ), '', sanitize_text_field( wp_unslash( $_SERVER['REQUEST_URI'] ) ) ) );

	/**
	 * Filter the request path before searching for a matching redirect.
	 *
	 * @param string $path
	 */
	$request_path = apply_filters( 'hm_redirects_request_path', $path );
	if ( ! $request_path ) {
		return false;
	}

	// Try to find a matching redirect.
	$redirect_uri = get_redirect_uri( $request_path );
	if ( ! $redirect_uri ) {
		return false;
	}

	/**
	 * Fires when the request path matches a redirect.
	 *
	 * @param string $redirect_uri Redirect-sanitised URL.
	 * @param string $request_path The request path that matched the redirect.
	 */
	do_action( 'hm_redirects_matched_redirect', $redirect_uri, $request_path );

	if ( ! wp_validate_redirect( $redirect_uri ) ) {
		return false;
	}

	header( 'X-HM-Redirects: true' );
	wp_safe_redirect( $redirect_uri, get_redirect_status_code( $request_path ) );
	exit;
}

/**
 * Retrieves the redirect URL.
 *
 * @param string $url The URL to redirect to.
 *
 * @return false|string Redirect-sanitised URL, or false if no valid redirect matched.
 */
function get_redirect_uri( $url ) {
	$url = Utilities\sanitise_and_normalise_url( $url );

	$redirect_post = get_redirect_post( $url );
	if ( is_null( $redirect_post ) ) {
		return false;
	}

	$to_url = $redirect_post->post_excerpt;
	if ( empty( $to_url ) ) {
		return false;
	}

	// If the URL is only a path, prefix it with the `home_url()`.
	$to_url = Utilities\prefix_path( $to_url );

	return wp_sanitize_redirect( $to_url );
}


/**
 * Retrieves the redirect status code.
 *
 * @param string $url The URL being redirected.
 *
 * @return int|string|WP_Error
 */
function get_redirect_status_code( $url = '' ) {
	if ( ! is_string( $url ) || strlen( $url ) === 0 ) {
		return new WP_Error( '', 'URL needs to be a non empty string' );
	}

	$url = Utilities\normalise_url( $url );
	if ( is_wp_error( $url ) ) {
		return 302;
	}

	$redirect_post = get_redirect_post( $url );

	if ( ! is_null( $redirect_post ) ) {
		$status_code = $redirect_post->post_content_filtered;
		if ( ! empty( $status_code ) ) {
			return $status_code;
		}
	}

	return 302;
}


/**
 * Retrieve the target URL's post ID.
 *
 * @param string $url The URL to retrieve the post for.
 *
 * @return int|null|string
 */
function get_redirect_post( $url ) {
	$url_hash      = Utilities\get_url_hash( $url );
	$redirect_post = get_page_by_path( $url_hash, OBJECT, [ Redirects_Post_Type\SLUG ] );
	if ( 'publish' !== get_post_status( $redirect_post ) ) {
		return null;
	}

	return $redirect_post;
}


includes/admin.php
<?php
/**
 * Adds a user interface for defining redirects.
 *
 * @package hm-redirects
 */

namespace HM\Redirects\Admin_UI;

use HM\Redirects\Post_Type as Redirects_Post_Type;
use HM\Redirects\Utilities;
use WP_Post;

/**
 * Register hooks.
 */
function setup() {
	add_action( 'add_meta_boxes', __NAMESPACE__ . '\\add_meta_box' );
	add_action( 'admin_enqueue_scripts', __NAMESPACE__ . '\\enqueue_scripts' );
	add_action( 'manage_' . Redirects_Post_Type\SLUG . '_posts_custom_column', __NAMESPACE__ . '\\posts_columns_content', 10, 2 );
	add_filter( 'manage_' . Redirects_Post_Type\SLUG . '_posts_columns', __NAMESPACE__ . '\\filter_posts_columns' );
	add_filter( 'post_row_actions', __NAMESPACE__ . '\\row_actions', 10, 2 );
	add_action( 'plugins_loaded', __NAMESPACE__ . '\\load_plugin_textdomain' );
	add_action( 'save_post', __NAMESPACE__ . '\\handle_redirect_saving', 13 );
}

/**
 * Add the metabox for the redirects.
 */
function add_meta_box() {
	\add_meta_box(
		'hm-redirects-meta',
		esc_html__( 'Redirect Settings', 'hm-redirects' ),
		__NAMESPACE__ . '\\output_meta_box',
		Redirects_Post_Type\SLUG,
		'normal',
		'high'
	);
}

/**
 * Enqueue scripts & styles.
 */
function enqueue_scripts() {
	if ( get_current_screen()->post_type !== Redirects_Post_Type\SLUG ) {
		return;
	}

	wp_enqueue_style( 'hm-redirects-admin', plugins_url( '/assets/admin.css', dirname( __FILE__ ) ), [], null, 'screen' );
}

/**
 * Filter the posts listing columns.
 *
 * @param array $columns List of column ids and labels.
 * @return array
 */
function filter_posts_columns( array $columns ) : array {
	$columns = [
		'title' => __( 'From', 'hm-redirects' ),
		'to' => __( 'To', 'hm-redirects' ),
		'status' => __( 'Status', 'hm-redirects' ),
		'date' => __( 'Date' ),
	];

	return $columns;
}

/**
 * Output for the custom admin columns.
 *
 * @param string $column Current column ID.
 * @param int    $post_id Current post ID.
 */
function posts_columns_content( string $column, int $post_id ) {
	$post = get_post( $post_id );

	if ( $column === 'to' ) {
		echo esc_html( $post->post_excerpt );
	}

	if ( $column === 'status' ) {
		echo intval( $post->post_content_filtered );
	}
}

/**
 * Add a new action to visit the redirected link.
 *
 * @param array   $actions Current row actions.
 * @param WP_Post $post Current row post.
 */
function row_actions( $actions, $post ) {
	if ( $post->post_status === 'trash' ) {
		return $actions;
	}

	$url = get_the_title( $post );
	/* translators: %s: Redirect source URL */
	$aria_label = esc_html( sprintf( __( 'Visit %s', 'hm-redirects' ), $url ) );
	return array_merge(
		[
			'visit' => sprintf(
				'<a target="_blank" href="%s" aria-label="%s">%s</a>',
				esc_url( $url ),
				$aria_label,
				esc_html__( 'Visit', 'hm-redirects' )
			),
		],
		$actions
	);
}

/**
 * Output the redirects metabox,
 *
 * @param WP_Post $post The currently edited post.
 */
function output_meta_box( WP_Post $post ) {
	$valid_status_codes = [ 301, 302, 303, 307, 403, 404 ];
	$status_code_labels = [
		301 => 'Moved Permanently',
		302 => 'Found',
		303 => 'See Other',
		307 => 'Temporary Redirect',
		403 => 'Forbidden',
		404 => 'Not Found',
	];

	/**
	 * Filter available status codes.
	 *
	 * @param array $status_code_labels Array of status codes and labels. The array keys are the status code and the values are the labels.
	 */
	$status_code_labels = apply_filters( 'hm_redirects_status_codes', $status_code_labels );

	/**
	 * Filter the default selected status code.
	 *
	 * @param int $default_status_code Defaults to 302.
	 */
	$default_status_code = apply_filters( 'hm_redirects_default_status_code', 302 );

	if ( ! in_array( $default_status_code, $valid_status_codes, true ) ) {
		$default_status_code = 302;
	}

	$status_code = ! empty( $post->post_content_filtered ) ? $post->post_content_filtered : $default_status_code;
	?>
	<p>
		<label for="hm_redirects_from_url"><?php esc_html_e( 'From URL', 'hm-redirects' ); ?></label><br>
		<input type="text" name="hm_redirects_from_url" id="hm_redirects_from_url" value="<?php echo esc_attr( $post->post_title ); ?>" class="code widefat"/>
	</p>
	<p class="description"><?php esc_html_e( 'This path should be relative to the root of the site.', 'hm-redirects' ); ?></p>

	<p>
		<label for="hm_redirects_to_url"><?php esc_html_e( 'To URL', 'hm-redirects' ); ?></label><br>
		<input type="text" name="hm_redirects_to_url" id="hm_redirects_to_url" value="<?php echo esc_attr( $post->post_excerpt ); ?>" class="code widefat"/>
	</p>

	<p>
		<label for="hm_redirects_status_code"><?php esc_html_e( 'HTTP Status Code:', 'hm-redirects' ); ?></label>
		<select name="hm_redirects_status_code" id="hm_redirects_status_code">
			<?php foreach ( $valid_status_codes as $code ) : ?>
				<option value="<?php echo esc_attr( $code ); ?>" <?php selected( $status_code, $code ); ?>><?php echo esc_html( $code . ' ' . $status_code_labels[ $code ] ); ?></option>
			<?php endforeach; ?>
		</select>
		<em><?php esc_html_e( "If you don't know what this is, leave it as is.", 'hm-redirects' ); ?></em>
	</p>
	<?php
	wp_nonce_field( 'hm_redirects', 'hm_redirects_nonce' );
}

/**
 * Save the redirect information.
 *
 * @param int $post_id Saved post id.
 *
 * @return bool Whether the redirect was saved successfully.
 */
function handle_redirect_saving( $post_id ) {
	if ( ! isset( $_POST['hm_redirects_nonce'] ) || ! wp_verify_nonce( sanitize_text_field( wp_unslash( $_POST['hm_redirects_nonce'] ) ), 'hm_redirects' ) ) {
		return false;
	}

	if ( wp_is_post_autosave( $post_id ) || wp_is_post_revision( $post_id ) ) {
		return false;
	}

	if ( ! isset( $_POST['hm_redirects_from_url'] ) || ! isset( $_POST['hm_redirects_to_url'] ) || ! isset( $_POST['hm_redirects_status_code'] ) ) {
		return false;
	}

	// phpcs:disable WordPress.VIP.ValidatedSanitizedInput
	// We're using a custom sanitisation function.
	$data = sanitise_and_normalise_redirect_data(
		wp_unslash( $_POST['hm_redirects_from_url'] ),
		wp_unslash( $_POST['hm_redirects_to_url'] ),
		wp_unslash( $_POST['hm_redirects_status_code'] )
	);
	// phpcs:enable

	$redirect_id = Utilities\insert_redirect( $data['from_url'], $data['to_url'], $data['status_code'], $post_id );

	return $redirect_id === $post_id;
}

/**
 * Sanitise and normalise data for a redirect post.
 *
 * @param string $unsafe_from        From URL.
 * @param string $unsafe_to          To URL.
 * @param string $unsafe_status_code HTTP status code.
 *
 * @return array
 */
function sanitise_and_normalise_redirect_data( $unsafe_from, $unsafe_to, $unsafe_status_code ) {
	return [
		'from_url'    => Utilities\normalise_url( Utilities\sanitise_and_normalise_url( $unsafe_from ) ),
		'to_url'      => Utilities\sanitise_and_normalise_url( $unsafe_to ),
		'status_code' => absint( $unsafe_status_code ),
	];
}

/**
 * Load the plugin translations.
 */
function load_plugin_textdomain() {
	\load_plugin_textdomain( 'hm-redirects', false, basename( dirname( dirname( __FILE__ ) ) ) . '/languages/' );
}
